File: packages/autonomo-ruby/test/test_autonomo.rb
#!/usr/bin/env ruby
# Test harness for autonomo-ruby
# Run: ruby test/test_autonomo.rb

require_relative '../lib/autonomo'

$passed = 0
$failed = 0

def test(name)
  yield
  puts "✅ #{name}"
  $passed += 1
rescue => e
  puts "❌ #{name}"
  puts "   #{e.message}"
  $failed += 1
end

def assert(condition, message = "Assertion failed")
  raise message unless condition
end

def assert_equal(expected, actual, message = nil)
  msg = message || "Expected #{expected.inspect}, got #{actual.inspect}"
  raise msg unless expected == actual
end

puts "\n🧪 autonomo-ruby Test Harness\n"

# Reset singletons
Autonomo::ElementRegistry.instance.clear

test "registry starts empty" do
  assert Autonomo::ElementRegistry.instance.list.empty?
end

test "register_tap_handler adds element" do
  tapped = false
  
  unregister = Autonomo.register_tap_handler("Test.Button") { tapped = true }
  
  assert Autonomo::ElementRegistry.instance.has?("Test.Button")
  assert Autonomo::ElementRegistry.instance.list.include?("Test.Button")
  
  # Invoke handler (takes value argument)
  handler = Autonomo::ElementRegistry.instance.get("Test.Button")
  handler.handler.call(nil)
  assert tapped, "Handler should have been called"
  
  unregister.call
  assert !Autonomo::ElementRegistry.instance.has?("Test.Button")
end

test "register_fill_handler works with value" do
  value = ""
  
  unregister = Autonomo.register_fill_handler(
    "Test.Input",
    get_value: -> { value }
  ) { |v| value = v }
  
  handler = Autonomo::ElementRegistry.instance.get("Test.Input")
  handler.handler.call("test value")
  assert_equal "test value", value
  assert_equal "test value", handler.get_value.call
  
  unregister.call
end

test "custom actions work" do
  unregister = Autonomo.register_custom_action("testAction") do |value|
    if value == "fail"
      Autonomo::ActionResult.fail("Intentional failure")
    else
      Autonomo::ActionResult.ok("Got: #{value}")
    end
  end
  
  result = Autonomo::CustomActionsRegistry.instance.execute("testAction", "hello")
  assert result.success, "Should succeed"
  assert result.message.include?("Got: hello")
  
  result = Autonomo::CustomActionsRegistry.instance.execute("testAction", "fail")
  assert !result.success, "Should fail"
  
  unregister.call
end

test "state manager tracks screen" do
  Autonomo::StateManager.instance.set_screen("login")
  assert_equal "login", Autonomo::StateManager.instance.get_screen
  
  st = Autonomo::StateManager.instance.get_state
  assert_equal "login", st.screen
end

test "state manager tracks user" do
  user = Autonomo::UserContext.new(id: "123", email: "test@example.com", role: "admin")
  Autonomo::StateManager.instance.set_user(user)
  
  st = Autonomo::StateManager.instance.get_state
  assert_equal "123", st.user.id
  assert_equal "test@example.com", st.user.email
end

test "state manager tracks errors" do
  Autonomo::StateManager.instance.clear_errors
  Autonomo::StateManager.instance.add_error("Test error")
  
  st = Autonomo::StateManager.instance.get_state
  assert st.errors.include?("Test error")
end

test "commands execute press" do
  pressed = false
  Autonomo.register_tap_handler("Cmd.Button") { pressed = true }
  
  result = Autonomo.execute_command("press", "Cmd.Button")
  assert result.success, "Command should succeed"
  assert pressed, "Button should be pressed"
  
  Autonomo::ElementRegistry.instance.unregister("Cmd.Button")
end

test "commands execute fill" do
  value = ""
  Autonomo.register_fill_handler("Cmd.Input") { |v| value = v }
  
  result = Autonomo.execute_command("fill", "Cmd.Input", "hello")
  assert result.success, "Command should succeed"
  assert_equal "hello", value
  
  Autonomo::ElementRegistry.instance.unregister("Cmd.Input")
end

test "commands return error for missing element" do
  result = Autonomo.execute_command("press", "Nonexistent.Button")
  assert !result.success, "Should fail"
  assert result.error.downcase.include?("not found")
end

puts "\n📊 Results: #{$passed} passed, #{$failed} failed\n"
exit($failed > 0 ? 1 : 0)

File: packages/autonomo-python/tests/test_autonomo.py
"""
Test harness for autonomo-python
Run: python -m pytest tests/test_autonomo.py -v
Or:  python tests/test_autonomo.py
"""

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from autonomo.registry import registry, register_tap_handler, register_fill_handler
from autonomo.actions import custom_actions, register_custom_action, ActionResult
from autonomo.state import state, UserContext
from autonomo.commands import execute_command

passed = 0
failed = 0

def test(name: str):
    def decorator(fn):
        global passed, failed
        try:
            fn()
            print(f"✅ {name}")
            passed += 1
        except AssertionError as e:
            print(f"❌ {name}")
            print(f"   {e}")
            failed += 1
        except Exception as e:
            print(f"❌ {name}")
            print(f"   {type(e).__name__}: {e}")
            failed += 1
    return decorator

print("\n🧪 autonomo-python Test Harness\n")

# Reset singletons
registry.clear()

@test("registry starts empty")
def _():
    assert len(registry.list()) == 0

@test("register_tap_handler adds element")
def _():
    tapped = {"value": False}
    
    def handler():
        tapped["value"] = True
    
    unregister = register_tap_handler("Test.Button", handler)
    
    assert registry.has("Test.Button")
    assert "Test.Button" in registry.list()
    
    # Invoke handler (takes Optional[str])
    elem = registry.get("Test.Button")
    elem.handler(None)
    assert tapped["value"] is True
    
    unregister()
    assert not registry.has("Test.Button")

@test("register_fill_handler works with value")
def _():
    value = {"current": ""}
    
    def handler(v):
        value["current"] = v
    
    unregister = register_fill_handler(
        "Test.Input",
        handler,
        get_value=lambda: value["current"]
    )
    
    elem = registry.get("Test.Input")
    elem.handler("test value")
    assert value["current"] == "test value"
    assert elem.get_value() == "test value"
    
    unregister()

@test("custom actions work")
def _():
    def my_action(value):
        if value == "fail":
            return ActionResult(success=False, error="Intentional failure")
        return ActionResult(success=True, message=f"Got: {value}")
    
    unregister = register_custom_action("testAction", my_action)
    
    result = custom_actions.execute("testAction", "hello")
    assert result.success is True
    assert "Got: hello" in result.message
    
    result = custom_actions.execute("testAction", "fail")
    assert result.success is False
    
    unregister()

@test("state manager tracks screen")
def _():
    state.set_screen("login")
    assert state.get_screen() == "login"
    
    snapshot = state.get_state()
    assert snapshot.screen == "login"

@test("state manager tracks user")
def _():
    user = UserContext(id="123", email="test@example.com", role="admin")
    state.set_user(user)
    
    snapshot = state.get_state()
    assert snapshot.user.id == "123"
    assert snapshot.user.email == "test@example.com"

@test("state manager tracks errors")
def _():
    state.clear_errors()
    state.add_error("Test error")
    
    snapshot = state.get_state()
    assert "Test error" in snapshot.errors

@test("commands execute press")
def _():
    pressed = {"value": False}
    register_tap_handler("Cmd.Button", lambda: pressed.__setitem__("value", True))
    
    result = execute_command("press", "Cmd.Button")
    assert result.success is True
    assert pressed["value"] is True
    
    registry.unregister("Cmd.Button")

@test("commands execute fill")
def _():
    value = {"current": ""}
    register_fill_handler("Cmd.Input", lambda v: value.__setitem__("current", v))
    
    result = execute_command("fill", "Cmd.Input", "hello")
    assert result.success is True
    assert value["current"] == "hello"
    
    registry.unregister("Cmd.Input")

@test("commands return error for missing element")
def _():
    result = execute_command("press", "Nonexistent.Button")
    assert result.success is False
    assert "not found" in result.error.lower()

@test("custom actions with metadata")
def _():
    from autonomo.actions import CustomActionMeta
    
    meta = CustomActionMeta(
        description="Greets the user",
        args={"name": "Name to greet"},
        example={"name": "World"}
    )
    
    unregister = register_custom_action(
        "greetAction",
        lambda value: ActionResult(success=True, message=f"Hello, {value}!"),
        meta=meta
    )
    
    # Verify action works
    result = custom_actions.execute("greetAction", "World")
    assert result.success is True
    assert "Hello, World!" in result.message
    
    # Verify get_all returns rich info
    all_actions = custom_actions.get_all()
    assert len(all_actions) >= 1
    
    greet_info = next((a for a in all_actions if a.name == "greetAction"), None)
    assert greet_info is not None
    assert greet_info.description == "Greets the user"
    assert greet_info.args == {"name": "Name to greet"}
    assert greet_info.example == {"name": "World"}
    
    unregister()

print(f"\n📊 Results: {passed} passed, {failed} failed\n")
sys.exit(1 if failed > 0 else 0)
